python/taichi/ad/_ad.py

```diff
@@ -205,24 +205,31 @@ def __enter__(self):
 
         # Attach the context manager to runtime
         self.runtime.target_tape = self
+        return self
 
     def __exit__(self, _type, value, tb):
         self.runtime.target_tape = None
         if self.eval_on_exit:
             self.grad()
+        if self.validation:
+            for f, _, mode_original in self.calls:
+                if mode_original == AutodiffMode.VALIDATION:
```

Comment:
OOC: why do we only revert to None when original is validation? (if we don't need revert others, we could choose save this mode for only a subset of kernels or simply assert on they are the same??

Reply:
It is because the validation only happens on forward kernels, i.e. kernels with None autodiff mode.